Plumb Azure interface's VPC / primary CIDR and set it as native routing CIDR in Azure IPAM mode

christarazi opened the pull request:
- azure, ipam, k8s: Derive primary / VPC CIDR of Azure interface
- daemon: Add Azure IPAM mode for setting the native routing CIDR

Related: https://github.com/cilium/cilium/pull/16672
Fixes: https://github.com/cilium/cilium/issues/16693

christarazi committed:
azure, ipam, k8s: Derive primary / VPC CIDR of Azure interface

To align with other CRD-backed IPAM modes such as ENI and Alibaba,
derive the VPC CIDR from the Azure API and set it as the native routing
CIDR. This enables the subsequent commit to use the CIDR to validate the
router IPs upon restoration.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

christarazi committed:
daemon: Add Azure IPAM mode for setting the native routing CIDR

This will allow the router IP restoration logic to pick up the correct
pod CIDR to validate the router IP.

This also fixes the issue where upon Cilium restart, additional IPs
were erroneously assigned to `cilium_host`.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

aanm commented on pkg/ipam/crd.go:
```diff
@@ -218,6 +218,14 @@ func deriveVpcCIDR(node *ciliumv2.CiliumNode) (result *cidr.CIDR) {
+			c, err := cidr.ParseCIDR(azif.CIDR)
+			if err == nil {
+				result = c
+			}
```
similarly to ENI, this should have a return here?

christarazi replied:
Fixed

aanm reviewed (approved):
LGTM, only a small nit

christarazi committed:
ipam: Fix return inside deriveVpcCIDR()

The `return` statement wasn't placed in the correct place, as the code
should return as soon as a valid result is found.

Signed-off-by: Chris Tarazi <chris@isovalent.com>
